Firestore/core/src/firebase/firestore/core/sync_engine.h

```diff
@@ -47,6 +46,44 @@ namespace firebase {
 namespace firestore {
 namespace core {
 
+/**
+ * Interface implemented by `EventManager` to handle notifications from
+ * `SyncEngine`.
+ */
+class QueryEventCallback {
+ public:
+  /** Handles a change in online state. */
+  virtual void HandleOnlineStateChange(model::OnlineState online_state) = 0;
+  /** Handles new view snapshots. */
+  virtual void OnViewSnapshots(std::vector<core::ViewSnapshot>&& snapshots) = 0;
+  /** Handles the failure of a query. */
+  virtual void OnError(const core::Query& query, util::Status error) = 0;
+
+  virtual ~QueryEventCallback() = default;
+};
+
+/**
+ * Interface implemented by `SyncEngine` receive requests from `EventManager`.
+ */
+class QueryEventSource {
+ public:
+  virtual void SetCallback(QueryEventCallback* callback) = 0;
+  /**
+   * Initiates a new listen. The FSTLocalStore will be queried for initial data
+   * and the listen will be sent to the `RemoteStore` to get remote data. The
+   * registered SyncEngineCallback will be notified of resulting view
+   * snapshots and/or listen errors.
+   *
+   * @return the target ID assigned to the query.
+   */
+  virtual model::TargetId Listen(Query query) = 0;
+
+  /** Stops listening to a query previously listened to via `Listen`. */
+  virtual void StopListening(const Query& query) = 0;
+
+  virtual ~QueryEventSource() = default;
```

Comment:
Destructor should be declared first within this class, according to the [style guide](https://google.github.io/styleguide/cppguide.html#Declaration_Order):
> _...generally prefer the following order: types (including typedef, using, and nested structs and classes), constants, factory functions, constructors, assignment operators, __destructor, all other methods,__ data members._

Applies to the other interface as well.